packages/common/src/audius-query/createRequestBatcher.ts

```diff
@@ -0,0 +1,140 @@
+  const requestGroups: Record<string, RequestGroup> = {}
+
+  const fetch =
+    <Args, Data>(
+      endpointName: string,
+      endpointConfig: EndpointConfig<Args, Data>
+    ) =>
+    (fetchArgs: Args, context: AudiusQueryContextType) => {
+      return new Promise<Data>((resolve, reject) => {
+        const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+
+        // Hash the request args and endpointName to determine the request group
+        const requestGroupKey = objectHash({
+          args: omit(fetchArgs ?? {}, idArgKey),
+          endpointName
+        })
+
+        requestGroups[requestGroupKey] = requestGroups[requestGroupKey] ?? {
+          queue: [],
+          timer: null
+        }
+
+        const requestGroup = requestGroups[requestGroupKey]
+
+        // Queue the request
+        requestGroup.queue = [
+          ...requestGroup.queue,
+          { fetchArgs, context, resolve, reject }
+        ]
+
+        // Set up a timer to perform the batch request
+        if (!requestGroup.timer) {
+          requestGroup.timer = setTimeout(
+            () => performBatch(requestGroup, endpointName, endpointConfig),
+            BATCH_PERIOD_MS
+          )
+        }
+      })
+    }
+
+  const performBatch = async <Args, Data>(
+    requestGroup: RequestGroup,
+    endpointName: string,
+    endpointConfig: EndpointConfig<Args, Data>
+  ) => {
+    const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+    const queuedRequests = [...requestGroup.queue] as RequestQueueItem<
+      Args,
+      Data
+    >[]
+
+    requestGroup.queue = []
+    requestGroup.timer = null
+
+    // Extract the IDs from the queued requests
+    const ids = queuedRequests.map(
+      ({ fetchArgs }) => fetchArgs[idArgKey]
+    ) as ID[]
+
+    // Use the context and other args from the most recent request
+    // Args are guaranteed to be the same for all requests in the group
+    // Context is not guaranteed to be the same, but we use the most recent one
+    const mostRecentRequest = queuedRequests[queuedRequests.length - 1]
+    const batchArgs = {
+      ...omit(mostRecentRequest.fetchArgs ?? {}, 'id'),
+      ids
+    } as { ids: ID[] } & Args
+
+    try {
+      // Make the batch request
+      const result = await endpointConfig.fetchBatch?.(
+        batchArgs,
+        mostRecentRequest.context
+      )
+
+      // Get the schema for the endpoint
+      const schema = schemas[endpointConfig.options.schemaKey!]
+
+      // This is necessary because the results are not guaranteed to be in the same order as the
+      // requested IDs
+      const resultsById = keyBy(result, (result) => {
+        return schema.getId(result)
+      })
+
+      queuedRequests.forEach((request) => {
+        if (result) {
+          const resultEntity =
+            resultsById[request.fetchArgs[endpointConfig.options.idArgKey!]]
+
+          if (resultEntity) {
+            request.resolve(resultEntity as any)
```

Comment:
I think this should work here?
```suggestion
            request.resolve(resultEntity as Data)
```

Reply:
Yeah that works, much nicer